R/time_bins.R

```diff
@@ -293,74 +304,124 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
   # Reorder dataframe
   df <- df[, c("bin", "interval_name", "rank", "max_ma", "mid_ma",
                "min_ma", "duration_myr", "abbr", "colour", "font")]
+  df <- df[order(-df$max_ma), ]
   # Generate equal-length bins? ------------------------------------------
-  if (is.numeric(size)) {
-      # Update bin size for age range
-      # How many bins should be generated?
-      n_bins <- round((max(df$max_ma) - min(df$min_ma)) / size)
-      size <- (max(df$max_ma) - min(df$min_ma)) / n_bins
-      #track cumulative sum
-      tracker <- list()
-      for (i in seq_len(length.out = nrow(df))) {
-        tracker[[i]] <- rep(NA, length.out = nrow(df))
-        tracker[[i]][i:nrow(df)] <- abs(
-          cumsum(df$duration_myr[i:nrow(df)]) - size)
-      }
-      # Calculate upper and lower limits for each bin
-      lower <- NULL
-      upper <- NULL
-      count <- 1
-      while (count <= nrow(df)) {
-        upper <- append(upper, which.min(tracker[[count]]))
-        lower <- append(lower, (count))
-        count <- which.min(tracker[[count]]) + 1
-      }
-      # Generate bin information
-      bin <- rev(seq_along(upper))
-      max_ma <- df[upper, c("max_ma")]
-      min_ma <- df[lower, c("min_ma")]
-      mid_ma <- (max_ma + min_ma) / 2
-      duration_myr <- max_ma - min_ma
-      intervals <- vector("character")
-      # Resolve any edge effect
-      if (duration_myr[length(duration_myr)] < size / 2) {
-        upper[length(upper) - 1] <- upper[length(upper)]
-        upper <- upper[-length(upper)]
-        lower <- lower[-length(lower)]
-        bin <- rev(seq_along(upper))
-        max_ma <- df[upper, c("max_ma")]
-        min_ma <- df[lower, c("min_ma")]
-        mid_ma <- (max_ma + min_ma) / 2
-        duration_myr <- max_ma - min_ma
+  if (is.numeric(size) || is.numeric(n_bins)) {
+    # Update bin size for age range
+    total_duration <- max(df$max_ma) - min(df$min_ma)
+    # How many bins should be generated?
+    if (is.numeric(size)) {
+      n_bins <- round(total_duration / size)
+      # Set n_bins to interval number if size results in too large n_bins
+      if (n_bins > nrow(df)) {
+          n_bins <- nrow(df)
+        }
+    } else {
+      # Throw error if n_bins is too large
+      if (n_bins > nrow(df)) {
+        stop("`n_bins` can't be greater than the number intervals.")
       }
-      # Get interval names
-      for (i in seq_along(upper)) {
-        intervals[i] <- toString(df[lower[i]:upper[i], c("interval_name")])
+    }
+    if (n_bins > 1) {
+    # updated bin size based on determined number of bins
```

Comment:
Everything in this `if` should be indented once

Reply:
Implemented